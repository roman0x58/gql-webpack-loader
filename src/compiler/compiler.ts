import path from 'path';
import webpack, { OutputFileSystem } from 'webpack';

import { createFsFromVolume, Volume } from 'memfs';
import { LoaderOptions } from "../LoaderOptions";

export default (fixture: string, options: LoaderOptions): Promise<[webpack.Stats, webpack.Compiler]> => {
    const compiler: webpack.Compiler = webpack({
        context: __dirname,
        entry: path.resolve(__dirname, `../../fixtures/${fixture}`),
        output: {
            path: path.resolve(__dirname),
            filename: 'bundle.js',
        },
        resolve: {
            extensions: ['.gql'],
            alias: {
                'gql-webpack-loader' : path.resolve(__dirname, '../../index.d.ts')
            }
        },
        module: {
            rules: [
                {
                    test: /\.gql$/,
                    use: [{
                        loader: path.resolve(__dirname, '../../index.js'),
                        options
                    }]
                }
            ],
        }
    });
    const webpackFs = createFsFromVolume(new Volume())
    compiler.outputFileSystem = webpackFs as unknown as OutputFileSystem;
    compiler.outputFileSystem.join = path.join.bind(path);
    (compiler as any).webpack.ModuleFilenameHelpers.matchObject = () => false

    return new Promise((resolve, reject) => {
        compiler.run((err, stats) => {
            if (err) reject(err);
            if (stats.hasErrors()) reject(stats.toJson().errors);
            resolve([stats, compiler]);
        });
    });
};
