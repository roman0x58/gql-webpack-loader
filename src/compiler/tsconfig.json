{
  "compilerOptions": {
    "esModuleInterop": true,
    "declaration": true
  },
  "exclude": ["./node_modules/"]
}