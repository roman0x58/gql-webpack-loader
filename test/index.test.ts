import compiler from '../src/compiler/compiler'
import webpack from "webpack";
import { IFs } from "memfs";
import path from 'path'

describe("gql loader", () => {
    test('transforms GraphQL query to TS module and generating appropriate declaration file using provided schema', async () => {
            return compiler('query.gql', {
                gqlSchemaPath: path.resolve(__dirname, '../fixtures/schema.ts'),
                mutationInterfaceName: 'MutationModel',
                queryInterfaceName: 'QueryModel',
                variablesSuffix: 'ArgModel'
            }).then(([stats, compiler]: [webpack.Stats, webpack.Compiler]) => {
                const output = stats.toJson({source: true});
                expect((compiler.outputFileSystem as unknown as IFs).readFileSync("./fixtures/query.gql.d.ts", "utf-8")).toMatchSnapshot('declaration')
                expect(output.modules[0].source).toMatchSnapshot('output')
            }).catch((errors) =>
                console.error(errors)
            )
        }, 15000
    )
})