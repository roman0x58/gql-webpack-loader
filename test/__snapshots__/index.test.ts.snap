// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gql loader transforms GraphQL query to ES6 module and generates appropriate declaration file using provided schema: declaration 1`] = `
"import type { GqlModule, GqlQuery } from 'gql-webpack-loader';
import type { QueryField3SomeQueryOperationArgsModel, QuerySomeQueryOperationArgsModel, MutationSomeMutationOperationArgsModel, QueryModel, MutationModel } from "./schema.ts";

export interface QueryOperationName1Query extends GqlQuery {
   "someQueryOperation"? : QueryModel["someQueryOperation"]    
}
export interface QueryOperationName2Query extends GqlQuery {
   "someQueryOperation"? : QueryModel["someQueryOperation"]    
}
export interface MutationOperationNameMutation extends GqlQuery {
   "someMutationOperation"? : MutationModel["someMutationOperation"]    
}

export declare const _default: {
	"queryOperationName1": GqlModule<QueryOperationName1Query, QueryField3SomeQueryOperationArgsModel & QuerySomeQueryOperationArgsModel>;
	"queryOperationName2": GqlModule<QueryOperationName2Query, QuerySomeQueryOperationArgsModel>;
	"mutationOperationName": GqlModule<MutationOperationNameMutation, MutationSomeMutationOperationArgsModel>
};

export default _default;"
`;

exports[`gql loader transforms GraphQL query to ES6 module and generates appropriate declaration file using provided schema: output 1`] = `
"export default {
	"queryOperationName1": { 
"query": \`query queryOperationName1($id: String, $x: Int) {
  someQueryOperation(id: $id) {
    field1
    field2
    field3(x: $x) {
      field3_1
    }
    ...Fragment1
  }
}
fragment Fragment1 on SomeOperationReturnType {
  field3
}\`,
"operation": "queryOperationName1" 
},
	"queryOperationName2": { 
"query": \`query queryOperationName2($id: String) {
  someQueryOperation(id: $id) {
    field1
  }
}\`,
"operation": "queryOperationName2" 
},
	"mutationOperationName": { 
"query": \`mutation mutationOperationName($id: String) {
  someMutationOperation(id: $id) {
    field1
    field2
  }
}\`,
"operation": "mutationOperationName" 
}
}"
`;

exports[`gql loader transforms GraphQL query to ES6 module with named import: declaration 1`] = `
"import type { GqlModule, GqlQuery } from 'gql-webpack-loader';
import type { QueryField3SomeQueryOperationArgsModel, QuerySomeQueryOperationArgsModel, MutationSomeMutationOperationArgsModel, QueryModel, MutationModel } from "./schema.ts";

export interface QueryOperationName1Query extends GqlQuery {
   "someQueryOperation"? : QueryModel["someQueryOperation"]    
}
export interface QueryOperationName2Query extends GqlQuery {
   "someQueryOperation"? : QueryModel["someQueryOperation"]    
}
export interface MutationOperationNameMutation extends GqlQuery {
   "someMutationOperation"? : MutationModel["someMutationOperation"]    
}

export declare const QueryGql: {
	"queryOperationName1": GqlModule<QueryOperationName1Query, QueryField3SomeQueryOperationArgsModel & QuerySomeQueryOperationArgsModel>;
	"queryOperationName2": GqlModule<QueryOperationName2Query, QuerySomeQueryOperationArgsModel>;
	"mutationOperationName": GqlModule<MutationOperationNameMutation, MutationSomeMutationOperationArgsModel>
};

export default QueryGql;"
`;

exports[`gql loader transforms GraphQL query to ES6 module with named import: named-import 1`] = `
"export const QueryGql = {
	"queryOperationName1": { 
"query": \`query queryOperationName1($id: String, $x: Int) {
  someQueryOperation(id: $id) {
    field1
    field2
    field3(x: $x) {
      field3_1
    }
    ...Fragment1
  }
}
fragment Fragment1 on SomeOperationReturnType {
  field3
}\`,
"operation": "queryOperationName1" 
},
	"queryOperationName2": { 
"query": \`query queryOperationName2($id: String) {
  someQueryOperation(id: $id) {
    field1
  }
}\`,
"operation": "queryOperationName2" 
},
	"mutationOperationName": { 
"query": \`mutation mutationOperationName($id: String) {
  someMutationOperation(id: $id) {
    field1
    field2
  }
}\`,
"operation": "mutationOperationName" 
}  
};
export default QueryGql;"
`;

exports[`gql loader transforms GraphQL query to ES6 module without declaration: output 1`] = `
"export default {
	"queryOperationName1": { 
"query": \`query queryOperationName1($id: String, $x: Int) {
  someQueryOperation(id: $id) {
    field1
    field2
    field3(x: $x) {
      field3_1
    }
    ...Fragment1
  }
}
fragment Fragment1 on SomeOperationReturnType {
  field3
}\`,
"operation": "queryOperationName1" 
},
	"queryOperationName2": { 
"query": \`query queryOperationName2($id: String) {
  someQueryOperation(id: $id) {
    field1
  }
}\`,
"operation": "queryOperationName2" 
},
	"mutationOperationName": { 
"query": \`mutation mutationOperationName($id: String) {
  someMutationOperation(id: $id) {
    field1
    field2
  }
}\`,
"operation": "mutationOperationName" 
}
}"
`;

exports[`gql loader transforms GraphQL query to ES6 module without variable interface name configuration: declaration 1`] = `
"import type { GqlModule, GqlQuery } from 'gql-webpack-loader';
import type { QueryField3SomeQueryOperationArgsModel, QuerySomeQueryOperationArgsModel, MutationSomeMutationOperationArgsModel, QueryModel, MutationModel } from "./schema.ts";

export interface QueryOperationName1Query extends GqlQuery {
   "someQueryOperation"? : QueryModel["someQueryOperation"]    
}
export interface QueryOperationName2Query extends GqlQuery {
   "someQueryOperation"? : QueryModel["someQueryOperation"]    
}
export interface MutationOperationNameMutation extends GqlQuery {
   "someMutationOperation"? : MutationModel["someMutationOperation"]    
}

export declare const _default: {
	"queryOperationName1": GqlModule<QueryOperationName1Query, QueryField3SomeQueryOperationArgsModel & QuerySomeQueryOperationArgsModel>;
	"queryOperationName2": GqlModule<QueryOperationName2Query, QuerySomeQueryOperationArgsModel>;
	"mutationOperationName": GqlModule<MutationOperationNameMutation, MutationSomeMutationOperationArgsModel>
};

export default _default;"
`;

exports[`gql loader transforms GraphQL query to ES6 module without variable interface name configuration: output 1`] = `
"export default {
	"queryOperationName1": { 
"query": \`query queryOperationName1($id: String, $x: Int) {
  someQueryOperation(id: $id) {
    field1
    field2
    field3(x: $x) {
      field3_1
    }
    ...Fragment1
  }
}
fragment Fragment1 on SomeOperationReturnType {
  field3
}\`,
"operation": "queryOperationName1" 
},
	"queryOperationName2": { 
"query": \`query queryOperationName2($id: String) {
  someQueryOperation(id: $id) {
    field1
  }
}\`,
"operation": "queryOperationName2" 
},
	"mutationOperationName": { 
"query": \`mutation mutationOperationName($id: String) {
  someMutationOperation(id: $id) {
    field1
    field2
  }
}\`,
"operation": "mutationOperationName" 
}
}"
`;
